src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp

```diff
@@ -82,315 +99,510 @@ enum FourHorsemen
     SAY_DEATH       = 4
 };
 
-class boss_four_horsemen : public CreatureScript
+struct boss_four_horsemen_baseAI : public BossAI
 {
-public:
-    boss_four_horsemen() : CreatureScript("boss_four_horsemen") { }
-
-    CreatureAI* GetAI(Creature* creature) const override
-    {
-        return GetInstanceAI<boss_four_horsemenAI>(creature);
-    }
-
-    struct boss_four_horsemenAI : public BossAI
-    {
-        boss_four_horsemenAI(Creature* creature) : BossAI(creature, BOSS_HORSEMEN)
+    public:
+        Creature* getHorsemanHandle(Horseman who) const
         {
-            Initialize();
-            id = Horsemen(0);
-            for (uint8 i = 0; i < 4; ++i)
-                if (me->GetEntry() == NPC_HORSEMEN[i])
-                    id = Horsemen(i);
-            caster = (id == HORSEMEN_LADY || id == HORSEMEN_SIR);
+            if (_which == who)
+                return me;
+            else
+                return ObjectAccessor::GetCreature(*me, instance->GetGuidData(who));
         }
-
-        void Initialize()
+        boss_four_horsemen_baseAI(Creature* creature, Horseman which, Position const* initialPath) : BossAI(creature, BOSS_HORSEMEN), _which(which), _initialPath(initialPath), _myMovementFinished(false), _nextMovement(0), _timeDied(0), _ourMovementFinished(false)
         {
-            uiEventStarterGUID.Clear();
-            nextWP = 0;
-            punishTimer = 2000;
-            nextMovementStarted = false;
-            movementCompleted = false;
-            movementStarted = false;
-            encounterActionAttack = false;
-            encounterActionReset = false;
-            doDelayPunish = false;
+            if (!me->IsAlive() && instance->GetBossState(BOSS_HORSEMEN) != DONE)
+                me->SetRespawnTime(10u);
```

Comment:
it shouldn't be needed to specify "u" after each number, please remove it here and in the next lines.